docs/specifications/linking-and-embedding-specification.md

```diff
@@ -0,0 +1,364 @@
+  *Also known as*: Index.
+- **Id**: Element attribute in the generated HTML that can be targeted with links / block embedding. Ids can also be used within the markdown files when defining anchors.
+- **Indexing**: Operation that goes through all the markdown files in the Foam workspace, finding all the links, block embeds, and link targets and finally linking these together in a Graph.
+- **Link Reference Definitions**: Information in a markdown file that converts wikilinks into regular markdown links.
+  *Also known as*: LinkRefDefs.
+  *See Also*: https://github.github.com/gfm/#link-reference-definitions
+- **Link**: A piece of text that points to a Target.
+- **Note**: A file containing text in markdown format.
+  *Also known as*: Foam Bubble, Document, Markdown file.
+- **Tag**: A piece of text that either acts as a link, or as an abstract categorization.
+- **Target**: An item (e.g. note, heading or an anchor) that can be pointed to with a link.
+- **Title**: The H1 heading of a note. A note can contain only a single title.
+- **Workspace**: The location (folder) where Foam Notes live in.
+
+## Targets
+
+The following items can be targetted via linking and / or embedding. The separation is explained in more detail in [Linking](#linking) and [Embedding](#embedding) chapters.
+
+### Notes
+
+Notes in Foam are individual markdown files. Notes can both reside in the Foam workspace root and any of its sub-folders. Both files and folders are encouraged to follow the kebab-case naming convention (e.g. `some/path-to/a-file.md`).
+
+### Headings
+
+Headings in Foam are denoted with markdown heading syntax, e.g. `## Heading on Second Level`. Each heading will have a corresponding id, a kebab-case version of the text (e.g. `heading-on-second-level`). In case there's multiple headings with the same text, their id is suffixed with a running number (e.g. `my-heading`, `my-heading-1`, `my-heading-2`).
+
+### Anchors
+
+User can add arbitrary anchors into a note with HTML-syntax: `<a name="my-anchor"></a>`. Since these names (identifiers) are written by users, they might end up being non-unique within a note. In these cases, an error is shown to the user. See the reasoning of using the name-attribute instead of the id-attribute from https://stackoverflow.com/a/7335259. In the future, better syntax for the anchors might be developed.
+
+### EBlocks
+
+While anchors define a single location in the document, eblocks define two; the beginning and the end of an eblock. The intent of a block is to specify content bounded by the begin and end markers. The end of an eblock can either be explicit or implicit.
+
+**Explicit eblocks** are defined with the anchor syntax and have two special rules; 1) the name of the beginning anchor starts with `begin:` and the ending anchor starts with `end:`, and 2) the other parts of the name must match. An example of matching anchors is:
+ 
+  ```markdown
+  <a name="begin:my-block"></a>
```

Comment:
Assuming we'll end up implementing explicit EBlocks, what about piggybacking on the fact that `name` attributes are not unique, and that html allows for [boolean attributes](https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#boolean-attributes) that don't define a value:

```html
<a begin name="my-block" />

Content here

<a end name="my-block />
```

It appears that the way browsers interpret this, upon clicking a link that references a duplicate named element, it always jumps to the first matched element. See: https://gist.github.com/jevakallio/eaa22f3038347cf2294eb9c75ef40ded#my-block

Again, I am aware that these are technically not valid HTML, but they could converted to be during build-time (or the end block could be dropped entirely, as it's only required during transclusion during edit and build time.

Given that the block end tag doesn't need to be valid (or even included, we could use more pseudo-HTML syntax:

```html
<a name="my-block" />

Content

<a close="my-block"  />
```

The exact wording here is up in the air, but would be keen to hear thoughts on using invalid HTML syntax, conceptually.

Reply:
Further, if we ended up implementing this as a syntax extension, such as MDX, what about the following:

```md
<block id="my-block">
  **Arbitrary** _markdown_ ~~content~~ allowed!
</block>
```

The MDX element could work as the link target boundaries during edit/build time, leave an a-tag, and pass through the content to the markdown processor so that it gets treated as if it was never embedded inside a HTML block at all:

```html
<a name="my-block">
<b>Arbitrary</b> <em>markdown</em> <strikethrough>content</strikethrough> allowed!
```

With this syntax, we would also not have to worry about clashing duplicate names or other arbitrary achor tags.

The drawback is that it does require a preprocessing step.